bee-message/src/payload/milestone/option/mod.rs

```diff
@@ -0,0 +1,230 @@
+    Pow(PowMilestoneOption),
+}
+
+impl MilestoneOption {
+    /// Return the milestone option kind of a [`MilestoneOption`].
+    pub fn kind(&self) -> u8 {
+        match self {
+            Self::Receipt(_) => ReceiptMilestoneOption::KIND,
+            Self::Pow(_) => PowMilestoneOption::KIND,
+        }
+    }
+}
+
+pub(crate) type MilestoneOptionCount = BoundedU8<0, { MilestoneOptions::COUNT_MAX }>;
+
+///
+#[derive(Clone, Debug, Eq, PartialEq, Deref, Packable)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[packable(unpack_error = Error, with = |e| e.unwrap_item_err_or_else(|p| Error::InvalidMilestoneOptionCount(p.into())))]
+pub struct MilestoneOptions(
+    #[packable(verify_with = verify_unique_sorted)] BoxedSlicePrefix<MilestoneOption, MilestoneOptionCount>,
+);
+
+impl TryFrom<Vec<MilestoneOption>> for MilestoneOptions {
+    type Error = Error;
+
+    #[inline(always)]
+    fn try_from(milestone_options: Vec<MilestoneOption>) -> Result<Self, Self::Error> {
+        Self::new(milestone_options)
+    }
+}
+
+impl IntoIterator for MilestoneOptions {
+    type Item = MilestoneOption;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        Vec::from(Into::<Box<[MilestoneOption]>>::into(self.0)).into_iter()
+    }
+}
+
+impl MilestoneOptions {
+    ///
+    pub const COUNT_MAX: u8 = 2;
+
+    /// Creates a new [`MilestoneOptions`].
+    pub fn new(milestone_options: Vec<MilestoneOption>) -> Result<Self, Error> {
+        let mut milestone_options =
+            BoxedSlicePrefix::<MilestoneOption, MilestoneOptionCount>::try_from(milestone_options.into_boxed_slice())
+                .map_err(Error::InvalidMilestoneOptionCount)?;
+
+        milestone_options.sort_by_key(MilestoneOption::kind);
+        // Sort is obviously fine now but uniqueness still needs to be checked.
+        verify_unique_sorted::<true>(&milestone_options)?;
+
+        Ok(Self(milestone_options))
+    }
+
+    /// Gets a reference to a [`MilestoneOption`] from a milestone option kind, if any.
+    #[inline(always)]
+    pub fn get(&self, key: u8) -> Option<&MilestoneOption> {
+        self.0
+            .binary_search_by_key(&key, MilestoneOption::kind)
+            // PANIC: indexation is fine since the index has been found.
+            .map(|index| &self.0[index])
+            .ok()
+    }
+
+    /// Gets a reference to a [`ReceiptMilestoneOption`], if any.
+    pub fn receipt(&self) -> Option<&ReceiptMilestoneOption> {
+        if let Some(MilestoneOption::Receipt(receipt)) = self.get(ReceiptMilestoneOption::KIND) {
+            Some(receipt)
+        } else {
+            None
+        }
+    }
+
+    /// Gets a reference to a [`PowMilestoneOption`], if any.
+    pub fn pow(&self) -> Option<&PowMilestoneOption> {
+        if let Some(MilestoneOption::Pow(pow)) = self.get(PowMilestoneOption::KIND) {
+            Some(pow)
+        } else {
+            None
+        }
```

Comment:
Does destructuring in closure arguments work just as it does for function args? Then you could do this:
`self.get(PowMilestoneOption::KIND).map(|MilestoneOption::Pow(pow)| pow)`

Reply:
It doesn't, this wouldn't be exhaustive

Reply:
Oh, yea, of course. My brain wanted to `map` this so badly that I didn't look at the enum properly.